### ISSUE TITLE
Dose composer support best checkpoint saver which can monitoring the checkpoint for best metrics or losses?

### ISSUE THREAD

**Issue by mayujie** [mosaicml/composer#2303]
## 🚀 Feature Request

From the Documentation in Checkpointing part, I didn't find there is a best checkpoint saver function supported. 
best checkpoint saver refers to the function that monitors the user-specified metrics or losses. It will save the best checkpoint for such particular best metrics or losses during the training process.
So I'm wondering if such a function already supports it or not.

In Keras, you can use the EarlyStopping function to achieve such needs. 
https://stackoverflow.com/questions/48285129/saving-best-model-in-keras

## Motivation

It would be convenient to have the best checkpoint saver function, in such case that I want to have the best checkpoint for a particular metric or loss. For example, for the Denoising task, I want to have the best checkpoint for PSNR metric and another best checkpoint for SSIM metric.

## [Optional] Implementation

CheckpointSaver class in composer.

**Comment by mvpatel2000** [comment]
This is a great suggestion!

We currently do have an EarlyStopping callback (see [docs](https://docs.mosaicml.com/projects/composer/en/latest/notes/early_stopping.html)). Unfortunately, at this time we don't have a way to save best checkpoint, but we will add it to our roadmap.

**Comment by ez2rok** [comment]
I had just mentioned this issue on the mosaic slack [here](https://mosaicml-community.slack.com/archives/C02H7NGT4T1/p1689728006155789) last week :).

**Comment by yasinalm** [comment]
is there any updates on this? this is a must-have feature.

**Comment by priba** [comment]
Hi @mvpatel2000, we have a custom best checkpoint saver but the issue is how to load the best checkpoint before evaluating, this means after the fit but before evaluation.

Using the tooling provided by Composer, it does not seem to work in DDP. We are currently using `checkpoint.load_checkpoint` and we have tested to load it in a single rank or broadcast the checkpoint to several ranks without luck. Wrapping the model with the `DistributedDataParallel`, changes the keymapping.

We have not check with deepspeed nor FSDP.

**Comment by mvpatel2000** [comment]
> Hi @mvpatel2000, we have a custom best checkpoint saver but the issue is how to load the best checkpoint before evaluating, this means after the fit but before evaluation.
> 
> Using the tooling provided by Composer, it does not seem to work in DDP. We are currently using `checkpoint.load_checkpoint` and we have tested to load it in a single rank or broadcast the checkpoint to several ranks without luck. Wrapping the model with the `DistributedDataParallel`, changes the keymapping.
> 
> We have not check with deepspeed nor FSDP.

@priba would you mind sharing a code example please? We're happy to take a look (and ideally we can add a unit test to ensure this works)

**Comment by priba** [comment]
Hi @mvpatel2000, thanks for your answer, here are some snippets of the code we are using.

```python
class BestCheckpointSaver(Callback):

    ...

    def fit_end(self, state: State, logger: Logger) -> None:
        if not self.save_last:
            return

        self.save_checkpoint("last" + DEFAULT_CHECKPOINT_EXTENSION, state, logger, None)

    def save_checkpoint(self, filename: str, state: State, logger: Logger, metric: Any) -> Path | None:
        destination = format_name_with_dist_and_time(
            os.fspath(Path(self.folder) / filename), state.run_name, state.timestamp
        )
        saved_path = checkpoint.save_checkpoint(state, destination, weights_only=self._weights_only)

        if saved_path is None:  # not all ranks save
            return None

        self.saved_checkpoints.append((Path(saved_path), state.timestamp.copy(), metric))
        log.debug(f"Uploading checkpoint to {destination} ...")
        logger.upload_file(remote_file_name=destination, file_path=saved_path)
        return Path(saved_path)
```

Then just before the eval call, we manually use a new entry point into the trainer:

```python
class Trainer:
    ...

    def load_best_checkpoint(self) -> None:
        """Loads the best checkpoint as tracked by BestCheckpointSaver.

        Example:
            ```
            trainer = Trainer(
                model=model,
                max_duration="100ep",
                callbacks=[BestCheckpointSaver("cross-entropy", mode="min")],
                ...
            )
            trainer.fit(...)
            trainer.load_best_checkpoint()
            trainer.eval(...)
            ```
        """

        if is_model_ddp(self.state.model) or is_model_deepspeed(self.state.model) or is_model_fsdp(self.state.model):
            log.warn(
                "`load_best_checkpoint` is not implemented for DDP, DeepSpeed and FSDP. "
                "Last weights will be used. Run test independently to load the best model."
            )
            return

        checkpoint_savers = [callback for callback in self.state.callbacks if isinstance(callback, BestCheckpointSaver)]

        if len(checkpoint_savers) == 0:
            raise ValueError("A `BestCheckpointSaver` callback must be provided to call this method.")

        if len(checkpoint_savers) > 1:
            log.warning(f"Several `BestCheckpointSaver` were provided. Evaluating with {checkpoint_savers[0]!r}.")

        ckpt = checkpoint_savers[0].best_checkpoint
        if ckpt is None:
            raise ValueError("No best checkpoint was found. Have you ran `.fit()` with an `eval_dataloader`?")

        checkpoint.load_checkpoint(os.fspath(ckpt[0]), self.state, self.logger)
```

We have tried different combinations such as running in rank 0, brodcasting the checkpoint path and even brodcasting the model weights but we didn't manage to make it work in DDP.

**Comment by mvpatel2000** [comment]
@priba looking into this... it would also be helpful if you could elaborate on what "it does not seem to work in DDP" means. Could you provide an error trace or some kind of description? This design seems more or less correct to me...

**Comment by priba** [comment]
Hi @mvpatel2000 , the issues we have are:

* As it is, ranks besides 0 will not have a checkpoint path and the Value error in `ckpt is None` will be raised.
* In case we just load the checkpoint in rank 0, it just runs into a deadlock
* In our case, we brodcasted the string `ckpt` to all ranks so we can load it everywhere, however, the keys of the state dict are not matching as DDP adds a `module.*`. I've seen that this should be removed inside the `load_checkpoint` but I am not sure why it is not happening.

**Comment by mvpatel2000** [comment]
@priba 
> * As it is, ranks besides 0 will not have a checkpoint path and the Value error in `ckpt is None` will be raised.
> * In case we just load the checkpoint in rank 0, it just runs into a deadlock
> * In our case, we brodcasted the string `ckpt` to all ranks so we can load it everywhere,
Only rank 0 saves the checkpoint to avoid duplicate work. But, all ranks need to load the checkpoint (with DDP, if you use FSDP we support a path for broadcasting). If you share a filesystem, broadcasting name should be sufficient as you described
> * however, the keys of the state dict are not matching as DDP adds a `module.*`. I've seen that this should be removed inside the `load_checkpoint` but I am not sure why it is not happening.
This is an ugly pytorch detail -- they add an extra level of `module.` when using DDP for reasons I am not familiar with. Composer should appropriately autostrip these on state_dict calls to `state`, but maybe this codepath fails. Could you share the exact error?

Torch 2.3 exposes a new API from torch `get_model_state_dict` which also handles this internally for PyTorch. So, when we release next week with 2.3 support and the new APIs, it might just go away for you

**Comment by priba** [comment]
Thanks for the info @mvpatel2000 

If we broadcast the string it does not throw any error message. It just skips loading the model as the keys are not there and throws a warning with the set of ignored keys. I don't think this is very informative for your case.

I will check once support for  2.3 is released and report back to you. Thank you so much

**Comment by mvpatel2000** [comment]
> If we broadcast the string it does not throw any error message. It just skips loading the model as the keys are not there and throws a warning with the set of ignored keys. I don't think this is very informative for your case.

Ah, I think this is a bug. If you load from `Trainer`, it will default raise an error if the keys are not there (`load_strict_model_weights` defaults to `True`). However, if you call `load_checkpoint` directly, it defaults to `False` whereas it should be consistent. I've opened a PR to fix this! https://github.com/mosaicml/composer/pull/3219
 
> I will check once support for 2.3 is released and report back to you. Thank you so much

Great! Please let me know if you encounter any errors after the release :)

**Comment by mvpatel2000** [comment]
@priba Hm... your snippet seems to work for me:
```
train          Epoch   0:  100%|_________________________| 25/25 [00:04<00:00,  6.76ba/s, loss/train/total=2.1885]                                                                                                                                  /mnt/workdisk/mihirp/composer/composer/core/data_spec.py:37: UserWarning: Cannot split tensor of length 904 into batches of size 1024. As it is smaller, no splitting will be done. This may happen on the last batch of a dataset if it is a smaller size than the microbatch size.
  warnings.warn(

eval           Epoch   0:  100%|_________________________| 5/5 [00:00<00:00,  6.62ba/s, metrics/eval/CrossEntropy=2.0713, metrics/eval/MulticlassAccuracy=0.2756]
train          Epoch   1:  100%|_________________________| 25/25 [00:03<00:00,  6.36ba/s, loss/train/total=1.9583]
eval           Epoch   1:  100%|_________________________| 5/5 [00:00<00:00,  6.92ba/s, metrics/eval/CrossEntropy=1.7521, metrics/eval/MulticlassAccuracy=0.3656]
eval           Epoch   0:  100%|_________________________| 5/5 [00:00<00:00,  7.07ba/s, metrics/eval/CrossEntropy=2.0718, metrics/eval/MulticlassAccuracy=0.2750]
```

Can you double check you are on Composer v0.22 and torch 2.3?

**Comment by mvpatel2000** [comment]
Great to hear! Please let me know if you encounter any further issues.

**Comment by mvpatel2000** [comment]
> @mvpatel2000 out of curiosity, may I ask you which GPUs did you use to run this test? Its much faster than what I got. Are these H100?

I was debugging this on 2xA100

**Comment by antoinebrl** [comment]
Hi @mvpatel2000 ! Thanks for providing feedback and insights into the the latest feature in PyTorch. 

We are also using 2xA100 with the same code snippet than shared above but our throughput is quite a bit lower, around ~4.1 ba/sec when you reach >6.6ba/sec. We are using A100 with 40GB of vRAM with CUDA 12.1 and the latest version of PyTorch. Are you using A100 with that amount of vRAM or GPUS with 80GB? Would you mind sharing the output of `composer_collect_env` so I can compare?
